pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -453,8 +494,100 @@ func (r *ReconcileConstraintTemplate) handleUpdate(
 		logger.Error(err, "error adding template to watch registry")
 		return reconcile.Result{}, err
 	}
+	// generating vap resources
+	if generateVap && constraint.IsVapAPIEnabled() {
+		currentVap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+		vapName := fmt.Sprintf("gatekeeper-%s", unversionedCT.GetName())
+		logger.Info("check if vap exists", "vapName", vapName)
+		if err := r.Get(ctx, types.NamespacedName{Name: vapName}, currentVap); err != nil {
+			if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
+				return reconcile.Result{}, err
+			}
+			currentVap = nil
+		}
+		logger.Info("get vap", "vapName", vapName, "currentVap", currentVap)
+		newVap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+		transformedVap, err := transform.TemplateToPolicyDefinition(unversionedCT)
+		if err != nil {
+			logger.Info("transform to vap error", "vapName", vapName, "error", err)
+			createErr := &v1beta1.CreateCRDError{Code: ErrCreateCode, Message: err.Error()}
+			status.Status.Errors = append(status.Status.Errors, createErr)
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not transform to vap object", status, err)
+			return reconcile.Result{}, err
+		}
+		if currentVap == nil {
+			newVap = transformedVap.DeepCopy()
+		} else {
+			newVap = currentVap.DeepCopy()
+			newVap.Spec = transformedVap.Spec
+		}
+
+		if err := controllerutil.SetControllerReference(ct, newVap, r.scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if currentVap == nil {
+			logger.Info("creating vap", "vapName", vapName)
+			if err := r.Create(ctx, newVap); err != nil {
+				logger.Info("creating vap error", "vapName", vapName, "error", err)
+				createErr := &v1beta1.CreateCRDError{Code: ErrCreateCode, Message: err.Error()}
+				status.Status.Errors = append(status.Status.Errors, createErr)
+				err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create vap object", status, err)
+				return reconcile.Result{}, err
+			}
+		}
+		if !reflect.DeepEqual(currentVap, newVap) {
```

Comment:
nit: I'd make this an else if statement, otherwise we make two writes to the API server instead of one (on account of currentVAP being nil, guaranteeing reflect.DeepEqual() will be false.